blockchain/stake/staketx.go

```diff
@@ -1221,3 +1222,73 @@ func FindSpentTicketsInBlock(block *wire.MsgBlock) *SpentTicketsInBlock {
 		RevokedTickets: revocations,
 	}
 }
+
+// CreateRevocationFromTicket creates a revocation transaction for the provided
+// ticket.
+func CreateRevocationFromTicket(ticketHash *chainhash.Hash,
+	ticketMinOuts []*MinimalOutput, revocationTxFee dcrutil.Amount,
+	revocationTxVersion uint16, params *chaincfg.Params) (*wire.MsgTx, error) {
+
+	// Create the revocation transaction and add the input.  The only input for a
+	// revocation transaction is a ticket submission (OP_SSTX tagged) output.
+	revocationTx := wire.NewMsgTx()
+	const ticketSubmissionOutput = 0
+	const tree = wire.TxTreeStake
+	ticketSubmissionAmount := ticketMinOuts[ticketSubmissionOutput].Value
+	prevOut := wire.NewOutPoint(ticketHash, ticketSubmissionOutput, tree)
+	txIn := wire.NewTxIn(prevOut, ticketSubmissionAmount, []byte{})
+	revocationTx.AddTxIn(txIn)
+
+	// Get the ticket commitment output details.
+	isP2SH, payToHashes, amounts, _, _, _ := SStxStakeOutputInfo(ticketMinOuts)
+
+	// Calculate the revocation output values from this data.
+	const revocationSubsidy = 0
+	revocationOutputAmounts := CalculateRewards(amounts, ticketSubmissionAmount,
+		revocationSubsidy)
+
+	// Add all the SSRtx-tagged transaction outputs to the transaction after
+	// performing some validity checks.
+	feeApplied := false
+	for i, payToHash := range payToHashes {
+		// Ensure amount is in the valid range for monetary amounts.
+		if amounts[i] <= 0 || amounts[i] > dcrutil.MaxAmount {
+			str := fmt.Sprintf("invalid SSTx amount: 0 >= %v > %v", amounts[i] <= 0,
+				dcrutil.MaxAmount)
+			return nil, stakeRuleError(ErrSStxBadCommitAmount, str)
+		}
+
+		// Create a new script which pays to the provided address specified in
+		// the original ticket tx.
+		var ssrtxOutScript []byte
+		switch isP2SH[i] {
+		case false: // P2PKH
+			addr, err := stdaddr.NewAddressPubKeyHashEcdsaSecp256k1V0(payToHash,
+				params)
+			if err != nil {
+				return nil, err
+			}
+			_, ssrtxOutScript = addr.PayRevokeCommitmentScript()
+		case true: // P2SH
+			addr, err := stdaddr.NewAddressScriptHashV0FromHash(payToHash, params)
+			if err != nil {
+				return nil, err
+			}
+			_, ssrtxOutScript = addr.PayRevokeCommitmentScript()
+		}
+
+		// Add the txout to our SSRtx tx.
+		amt := revocationOutputAmounts[i]
+		if !feeApplied && int64(revocationTxFee) < amt {
+			amt -= int64(revocationTxFee)
+			feeApplied = true
+		}
+		txOut := wire.NewTxOut(amt, ssrtxOutScript)
+		revocationTx.AddTxOut(txOut)
+	}
+
+	// Set the version of the revocation transaction.
+	revocationTx.Version = revocationTxVersion
+
```

Comment:
Maybe also sanity check that the sum of committed amounts equal the input amount? This is protected elsewhere in consensus code and thus doing it here is just ensuring the inputs to this function are valid (so when this ends up as part of consensus the inputs should be sane).

Reply:
This function already calls `SStxStakeOutputInfo` to get the input amounts and `CalculateRewards` to calculate the output amounts using those input amounts, so I think it should be the responsibility of those functions to ensure they are returning correct results rather than re-checking again here.